bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1898,302 +1788,81 @@ void handleBookieFailure(final Map<Integer, BookieSocketAddress> failedBookies)
-            writeLedgerConfig(new ChangeEnsembleCb(ensembleInfo, curBlockAddCompletions,
-                    ensembleChangeIdx));
-            return true;
+            handleUnrecoverableErrorDuringAdd(WriteException);
+            return;
         }
 
-        @Override
-        public String toString() {
-            return String.format("ReReadLedgerMetadata(%d)", ledgerId);
-        }
+        new MetadataUpdateLoop(
+                clientCtx.getLedgerManager(), getId(),
+                this::getLedgerMetadata,
+                (metadata) -> {
+                    List<BookieSocketAddress> lastEnsemble = metadata.getLastEnsembleValue();
+                    boolean failedBookieInEnsemble = failedBookies.entrySet().stream()
+                        .anyMatch((e) -> lastEnsemble.get(e.getKey()).equals(e.getValue()));
+                    return !metadata.isClosed() && !metadata.isInRecovery() && failedBookieInEnsemble;
+                },
+                (metadata) -> {
+                    List<BookieSocketAddress> newEnsemble = EnsembleUtils.replaceBookiesInEnsemble(
+                            clientCtx.getBookieWatcher(), getId(), metadata, getCurrentEnsemble(), failedBookies);
+                    Long lastEnsembleKey = metadata.getLastEnsembleKey();
+                    LedgerMetadataBuilder builder = LedgerMetadataBuilder.from(metadata);
+                    long newEnsembleStartEntry = getLastAddConfirmed() + 1;
+                    checkState(lastEnsembleKey <= newEnsembleStartEntry,
+                               "New ensemble must either replace the last ensemble, or add a new one");
+                    if (lastEnsembleKey.equals(newEnsembleStartEntry)) {
+                        return builder.replaceEnsembleEntry(newEnsembleStartEntry, newEnsemble).build();
+                    } else {
+                        return builder.newEnsembleEntry(newEnsembleStartEntry, newEnsemble).build();
+                    }
+                },
+                this::setLedgerMetadata)
+            .run().whenCompleteAsync((metadata, ex) -> {
+                    if (ex != null) {
+                        handleUnrecoverableErrorDuringAdd(BKException.getExceptionCode(ex, WriteException));
+                    } else if (metadata.isClosed()) {
+                        handleUnrecoverableErrorDuringAdd(BKException.Code.LedgerClosedException);
+                    } else if (metadata.isInRecovery()) {
+                        handleUnrecoverableErrorDuringAdd(BKException.Code.LedgerFencedException);
+                    } else {
+                        synchronized (metadataLock) {
+                            if (!delayedWriteFailedBookies.isEmpty()) {
+                                Map<Integer, BookieSocketAddress> toReplace = new HashMap<>(delayedWriteFailedBookies);
+                                delayedWriteFailedBookies.clear();
+
+                                ensembleChangeLoop(origEnsemble, toReplace);
```

Comment:
At this point, setLedgerMetadata() must have updated the current metadata to the updated value. Why are you restarting the loop with origEnsemble?

Reply:
origEnsemble is used once all ensembleChange have occurred to check which indices have changed, and then resend the pending adds to them.